pennylane/devices/default_clifford.py

```diff
@@ -0,0 +1,644 @@
+
+    def measure(self, circuit, tableau_simulator, global_phase, stim):
+        """Given a circuit, compute and return the measurement results."""
+        results = []
+        for meas in circuit.measurements:
+            # Analytic case
+            if not circuit.shots:
+                # Computing density matrix via tableaus
+                if isinstance(meas, DensityMatrixMP):  # do first because it is a child of StateMP
+                    res = self._measure_density_matrix(tableau_simulator)
+
+                # Computing statevector via tableaus
+                elif isinstance(meas, StateMP):
+                    res = self._measure_state(tableau_simulator, circuit, global_phase)
+
+                # Computing purity via tableaus
+                elif isinstance(meas, PurityMP):
+                    res = self._measure_purity(meas, circuit)
+
+                # Computing expectation values via measurement
+                elif isinstance(meas, ExpectationMP):
+                    res = self._measure_expectation(tableau_simulator, meas, stim)
+
+                # Computing more measurements
+                else:
+                    raise NotImplementedError(
+                        f"default.clifford doesn't support the {type(meas)} measurement at the moment."
+                    )
+
+                results.append(res)
+
+        return results[0] if len(results) == 1 else tuple(results)
+
+    def _measure_state(self, tableau_simulator, circuit, global_phase):
+        """Measure the state of the simualtor device."""
+        if self._tableau:
+            tableau = tableau_simulator.current_inverse_tableau().inverse()
+            x2x, x2z, z2x, z2z, x_signs, z_signs = tableau.to_numpy()
+            pl_tableau = np.vstack(
+                (
+                    np.hstack((x2x, x2z, x_signs.reshape(-1, 1))),
+                    np.hstack((z2x, z2z, z_signs.reshape(-1, 1))),
+                )
+            ).astype(int)
+            if pl_tableau.shape == (0, 1) and circuit.num_wires:
+                return np.array([[1.0, 0.0, 0.0], [0.0, 1.0, 0.0]])
+            return pl_tableau
+
+        state = qml.math.array(tableau_simulator.state_vector(endian="big"))
+        if state.shape == (1,) and circuit.num_wires:
+            # following is faster than using np.eye(length=1, size, index)
+            state = qml.math.zeros(2**circuit.num_wires, dtype=complex)
+            state[0] = 1.0 + 0.0j
+        return state * qml.matrix(global_phase)
+
+    @staticmethod
+    def _measure_density_matrix(tableau_simulator):
+        """Measure the density matrix from the state of simulator device."""
+        state_vector = qml.math.array(tableau_simulator.state_vector(endian="big"))
+        return qml.math.einsum("i, j->ij", state_vector, state_vector)
+
+    @staticmethod
+    def _measure_purity(meas_op, circuit):
+        """Measure the purity of the state of simulator device"""
+        if circuit.op_wires != meas_op.wires:
```

Comment:
the `op_wires` property is similarly hacky, and came with `validate_device_wires` - you probably just want `circuit.wires`. For example:
```pycon
>>> qs = qml.tape.QuantumScript(measurements=[qml.expval(qml.PauliZ(0))])
>>> qs.wires, qs.op_wires
(<Wires = [0]>, <Wires = []>)
```
The purity measurement would probably be valid here, but this would cause a `NotImplementedError`

Reply:
Added it. Will change it in the second PR as well.